nca/FWRules/InteractiveConnectivityGraph.py

```diff
@@ -0,0 +1,547 @@
+            """
+            adding to the element_relation of t_id the elements that should be on every svg file
+            param: t_id: string
+            param: element_relation: ElementRelations
+            """
+            element_relation.relations.add(t_id)
+            element_relation.highlights.add(t_id)
+            element_relation.relations.add(self.graph.t_id)
+            for conn in self.graph.conn_legend.conns.values():
+                element_relation.relations.add(conn.t_id)
+            # to remain all pods in all graphs:
+            # elements_relations[t_id].relations |= set(n.t_id for n in self.graph.nodes.values() if n.real_node())
+
+        def set_tags_relations(self):
+            """
+            find the related elements of each element
+            return:  elements_relations dict {str: ElementRelations}: for each element list of relations and list of highlights
+            """
+            elements_relations = defaultdict(InteractiveConnectivityGraph.ElementRelations)
+
+            # first add the basic relations:
+            all_items = list(self.graph.conn_legend.conns.values()) + list(self.graph.edges.values()) +\
+                list(self.graph.nodes.values()) + list(self.graph.namespaces.values()) + [self.graph]
+            for item in all_items:
+                self._add_basic_relations(item.t_id, elements_relations[item.t_id])
+
+            # the grapg itself will have all the items:
+            elements_relations[self.graph.t_id].relations |= set(item.t_id for item in all_items)
+
+            # add namespaces to nodes:
+            for namespace in self.graph.namespaces.values():
+                for node in namespace.nodes:
+                    elements_relations[node.t_id].relations.add(namespace.t_id)
+
+            # add nodes of edge to an edge
+            # add edge to its connectivity
+            for edge in self.graph.edges.values():
+                elements_relations[edge.t_id].relations |= elements_relations[edge.src.t_id].relations
+                elements_relations[edge.t_id].relations |= elements_relations[edge.dst.t_id].relations
+                elements_relations[edge.conn.t_id].relations |= elements_relations[edge.t_id].relations
+
+            # add to nodes its edges
+            for node in self.graph.nodes.values():
+                for edge in node.edges:
+                    elements_relations[node.t_id].relations |= elements_relations[edge.t_id].relations
+
+            # add edge and nodes to all clique elements (conns already inside)
+            # highlights all clique core elements with each other
+            for clique in self.graph.cliques:
+                for el in clique.nodes + clique.edges:
+                    for e in clique.edges:
+                        elements_relations[el.t_id].relations |= elements_relations[e.t_id].relations
+                        elements_relations[clique.conn.t_id].relations |= elements_relations[e.t_id].relations
```

Comment:
Seems like we are adding to `elements_relations[clique.conn.t_id].relations` the same sets of edge relations over and over again. Can this be improved?

Reply:
done